### ISSUE TITLE
how to change `this.foo(a, b)` to `this._lookup('foo', [a, b])`?

### ISSUE THREAD

**Issue by faceyspacey** [thejameskyle/babel-plugin-handbook#20]
I'm looking to do some basic transpilations of my jsx code. How would you accomplish something like that? 

Another simple transpilation I'm looking to make is:

`<MyComponent />`
to:
`<MyComponent parent={this} />`

Perhaps simple *regex/replace* style examples like these would make for a good section of the handbook as well. The problem of course with regexes is that they can't handle recursive replaces well. For example:

`this.foo(this.bar())`
 would become:
`this.lookup('foo', [this.bar(]));`

I imagine a lot of people may have similar challenges where they are stuck with regexes and need to upgrade to the next more powerful thing, i.e. a full on parser.

**Comment by thejameskyle** [comment]
I'm not sure what you are asking for

**Comment by faceyspacey** [comment]
I'm just stating a specific challenge I'm looking to figure out, and explaining how it might be a common problem a lot of your users or potential users might have. Specifically: recursively replacing code based on a regex, or of a certain identifier type, etc.

**Comment by thejameskyle** [comment]
It's generally a really bad idea to use regex for code transformation.

Recursion in Babel (or any compiler) just works though, it's just the nature of traversal.

**Comment by faceyspacey** [comment]
Exactly. ..I just updated my initial post, as I had left out some information about the code transformation I'm looking to do. ..Basically, the sentiment you expressed "it just works" is the idea I--as a code parsing beginning--get. So it seems that a basic example to help people like me "level up" from using regexes to creating a Babel plugin would make a lot of sense.

**Comment by thejameskyle** [comment]
So you want an example where I take some code being transformed with regex and turn it into a Babel plugin?

**Comment by faceyspacey** [comment]
Yea, particularly code that would break because of regexes' lacking in the recursion department. 

My exact example may in fact be the canonical example here. Here it is again succinctly:

```
let reg = /(?:this\.(\w+)\(([\s\S]*?)\))/g;
let subst = 'this.lookup(\'$1\', [$2])'; 
let str = 'this.foo(this.bar());';

let result = str.replace(reg, subst);
```

We are hoping for this recursive result:
```
this.lookup('foo', [this.lookup('bar', [])]); //notice this.lookup() called within the initial substitution
```
but in regexes only get:
```
this.lookup('foo', [this.bar(]))
```

**Comment by thejameskyle** [comment]
I would have to think of a more relatable example.

**Comment by faceyspacey** [comment]
ok

**Comment by fab1an** [comment]
Maybe one could create a helper-function which checks that an ast-subtree is shaped in a certain way, like

    memberexpression
    - ident ("obj")
    - ident ("property)

allows a (regexp?) transformation on a string-representation `obj.property` and rebuilds the ast-subtree?

I agree with @thejameskyle  though, it's better to not use regular expressions. They are harder to grasp as well, don't you think?

**Comment by faceyspacey** [comment]
@fab1an well the whole point is to not use regular expressions because they simply can't do the stuff you can do with access to the AST. Nobody's arguing that. But many more people know how to work with regular expressions than ASTs--so the goal is to best bridge them over.

**Comment by fab1an** [comment]
I'm still not sure what you are trying to achieve.

You can already work with regular expressions on node names and the like. If you want to transform an ast with a regular expression you would first have to turn the ast back into a string-representation (effectively "unparsing" it) and then work on that.

That should already work: just read the code via `babylon.parse` and recursively perform your regexp on it?

**Comment by faceyspacey** [comment]
im trying to achieve *NOT* using regexes anymore. here's almost exactly what i was trying to achieve:

http://stackoverflow.com/questions/34100079/need-regex-to-accomplish-basic-recursive-feature-of-language-parsers-or-help-ma/34109901#34109901

Basically I'm looking to learn more about how ASTs work and how babel's plugin API works. The idea is to make it easy for people like me to learn how to do things using Babel's AST API that you were trying to achieve via regexes and failing.

**Comment by fab1an** [comment]
So you want a code-example WITHOUT a regexp that replaces sample code that used a regexp? This is what the stackoverflow answer already includes, isn't it?

Or you want a code-example that does _generic_ transformation using supplied regular expressions? That's not possible, i'm afraid, without detailed specification of how it should work.

**Comment by faceyspacey** [comment]
the first one. The stackoverflow answer is almost exactly it. Basically it would be nice if the docs had an in depth version of that for many similar use cases, while also introducing you to the API's methods. This is the "guide" approach, which is the reverse of what there already is, which is leading with the API. So basically in another section of the handbook, lead with a few common tasks you might wanna do.
